docs/dev/how-to/send-transaction-l1-l2.md

```diff
@@ -0,0 +1,334 @@
+
+## Common use cases
+
+Along with zkSync Era's built-in censorship resistance that requires multi-layer interoperability, there are some common use cases that need L1 to L2 transaction functionality, such as:
+
+- Custom bridges.
+- Multi-layer governing smart contracts.
+
+## Step-by-step
+
+1. Import the zkSync Era library or contract containing the required functionality. 
+
+    The import gives access to the [`IZkSync.sol`](https://github.com/matter-labs/v2-testnet-contracts/blob/b8449bf9c819098cc8bfee0549ff5094456be51d/l1/contracts/zksync/interfaces/IZkSync.sol#L4) inherited interfaces that include the gas estimation functionality.
+
+    Import the contracts with yarn (recommended), or [download the contracts](https://github.com/matter-labs/v2-testnet-contracts) from the repo.
+
+    ::: code-tabs
+    @tab yarn
+    ```yarn
+    yarn add -D @matterlabs/zksync-contracts
+    ```
+    :::
+
+2. Get the current L1 gas price with Ethereum JSON-RPC method [`eth_gasPrice`](https://ethereum.github.io/execution-apis/api-documentation/).
+
+    ::: code-tabs
+    @tab TypeScript
+    ```ts
+    // Method available on a zkSync Era JS SDK Provider object
+    override async getGasPrice(token?: Address): Promise<BigNumber> {
+        const params = token ? [token] : [];
+        const price = await this.send('eth_gasPrice', params);
+        return BigNumber.from(price);
+    }
+    ```
+    :::
+
+3. Call the JSON-RPC method [`zks_estimateGasL1toL2`](../../api/api.md#zks-estimategasl1tol2), wrapping the transaction data in a [`CallRequest`](#transaction-parameters-1) JSON object parameter. 
+
+    The method returns the amount of gas required for the transaction to succeed. 
+
+    :::tip Important
+    This value is often referred to as **limit, or gas limit, or L2 gas limit** in our documented examples. 
+    :::
+
+    ::: code-tabs
+    @tab TypeScript
+    ```ts
+    // available on a zkSync Era JS SDK Provider object
+    async estimateGasL1(transaction: utils.Deferrable<TransactionRequest>): Promise<BigNumber> {
+        await this.getNetwork();
+        const params = await utils.resolveProperties({
+            transaction: this._getTransactionRequest(transaction)
+        });
+        if (transaction.customData != null) {
+            // @ts-ignore
+            params.transaction.customData = transaction.customData;
+        }
+        const result = await this.send('zks_estimateGasL1ToL2', [
+            Provider.hexlifyTransaction(params.transaction, { from: true })
+        ]);
+        try {
+            return BigNumber.from(result);
+        } catch (error) {
+            throw new Error(`bad result from backend (zks_estimateGasL1ToL2): ${result}`);
+        }
+    }
+    ```
+    :::
+
+    3.1 Apply an alias to the addresses in the request if the sender address is a contract. 
```

Comment:
the alias should be applied to the address of the sender before the estimation is done (i.e. the steps 3.1 and 3.0 should be swapped)

Reply:
moved around and renumbered